core/logger/logger.go

```diff
@@ -122,31 +128,68 @@ func newProductionConfig(dir string, jsonConsole bool, toDisk bool, unixTS bool)
 	return config
 }
 
-type Config interface {
-	LogFileDir() string
-	JSONConsole() bool
-	LogToDisk() bool
-	LogLevel() zapcore.Level
-	LogUnixTimestamps() bool
+// NewLogger returns a new Logger configured from environment variables, and logs any parsing errors.
+// Tests should use TestLogger.
+func NewLogger() Logger {
+	var c Config
+	var parseErrs []string
+
+	var invalid string
+	c.LogLevel, invalid = envvar.LogLevel.ParseLogLevel()
+	if invalid != "" {
+		parseErrs = append(parseErrs, invalid)
+	}
+
+	c.Dir = os.Getenv("LOG_FILE_DIR")
+	if c.Dir == "" {
+		var invalid2 string
+		c.Dir, invalid2 = envvar.RootDir.ParseString()
+		if invalid2 != "" {
+			parseErrs = append(parseErrs, invalid2)
+		}
+	}
+
+	c.JsonConsole, invalid = envvar.JSONConsole.ParseBool()
+	if invalid != "" {
+		parseErrs = append(parseErrs, invalid)
+	}
+
+	c.ToDisk, invalid = envvar.LogToDisk.ParseBool()
+	if invalid != "" {
+		parseErrs = append(parseErrs, invalid)
+	}
+
+	c.UnixTS, invalid = envvar.LogUnixTS.ParseBool()
+	if invalid != "" {
+		parseErrs = append(parseErrs, invalid)
+	}
+
+	l := c.New()
+	for _, msg := range parseErrs {
+		l.Error(msg)
+	}
+	return l
 }
 
-// NewLogger returns a new Logger configured by c with pretty printing to stdout.
-// If LogToDisk is false, the Logger will only log to stdout.
-// Tests should use TestLogger instead.
-func NewLogger(c Config) Logger {
-	return newLogger(c.LogLevel(), c.LogFileDir(), c.JSONConsole(), c.LogToDisk(), c.LogUnixTimestamps())
+type Config struct {
+	LogLevel    zapcore.Level
+	Dir         string
+	JsonConsole bool
+	ToDisk      bool // if false, the Logger will only log to stdout.
+	UnixTS      bool
 }
 
-// newLogger returns a new production Logger with sentry forwarding.
-func newLogger(logLevel zapcore.Level, dir string, jsonConsole bool, toDisk bool, unixTS bool) Logger {
-	cfg := newProductionConfig(dir, jsonConsole, toDisk, unixTS)
-	cfg.Level.SetLevel(logLevel)
+// New returns a new Logger with pretty printing to stdout, prometeus counters, and sentry forwarding.
+// Tests should use TestLogger.
+func (c *Config) New() Logger {
```

Comment:
Could we not have the config just store the logger on itself, then return this when you call new?

That way we don't have two ways of parsing the env vars to create a logger.

Reply:
nvm I misunderstood the code, this looks good